Tests/CarthageKitTests/ValidateSpec.swift

```diff
@@ -0,0 +1,169 @@
+@testable import CarthageKit
+import Foundation
+import Nimble
+import Quick
+import ReactiveSwift
+import Result
+import Tentacle
+
+private extension CarthageError {
+	var compatibilityInfos: [CompatibilityInfo] {
+		if case let .invalidResolvedCartfile(infos) = self {
+			return infos
+		}
+		return []
+	}
+}
+
+class ValidateSpec: QuickSpec {
+	override func spec() {
+		let validCartfile = """
+					github "Alamofire/Alamofire" "4.6.0"
+					github "CocoaLumberjack/CocoaLumberjack" "3.4.1"
+					github "Moya/Moya" "10.0.2"
+					github "ReactiveCocoa/ReactiveSwift" "2.0.1"
+					github "ReactiveX/RxSwift" "4.1.2"
+					github "antitypical/Result" "3.2.4"
+					github "yapstudios/YapDatabase" "3.0.2"
+					"""
+
+		let invalidCartfile = """
+					github "Alamofire/Alamofire" "5.0.0"
+					github "CocoaLumberjack/CocoaLumberjack" "commitish"
+					github "Moya/Moya" "10.0.2"
+					github "ReactiveCocoa/ReactiveSwift" "2.0.1"
+					github "ReactiveX/RxSwift" "4.1.2"
+					github "antitypical/Result" "4.0.0"
+					github "yapstudios/YapDatabase" "3.0.2"
+					"""
+
+		let moyaDependency = Dependency.gitHub(.dotCom, Repository(owner: "Moya", name: "Moya"))
+		let resultDependency = Dependency.gitHub(.dotCom, Repository(owner: "antitypical", name: "Result"))
+		let alamofireDependency = Dependency.gitHub(.dotCom, Repository(owner: "Alamofire", name: "Alamofire"))
+		let reactiveSwiftDependency = Dependency.gitHub(.dotCom, Repository(owner: "ReactiveCocoa", name: "ReactiveSwift"))
+		let rxSwiftDependency = Dependency.gitHub(.dotCom, Repository(owner: "ReactiveX", name: "RxSwift"))
+		let yapDatabaseDependency = Dependency.gitHub(.dotCom, Repository(owner: "yapstudios", name: "YapDatabase"))
+		let cocoaLumberjackDependency = Dependency.gitHub(.dotCom, Repository(owner: "CocoaLumberjack", name: "CocoaLumberjack"))
+
+		// These tuples represent the desired version of a dependency, paired with its parent dependency;
+		// moya_3_1_0 indicates that Moya expects a version compatible with 3.1.0 of *another* dependency
+		let moya_3_1_0 = (moyaDependency, VersionSpecifier.compatibleWith(SemanticVersion(major: 3, minor: 1, patch: 0)))
+		let moya_4_1_0 = (moyaDependency, VersionSpecifier.compatibleWith(SemanticVersion(major: 4, minor: 1, patch: 0)))
+		let reactiveSwift_3_2_1 = (reactiveSwiftDependency, VersionSpecifier.compatibleWith(SemanticVersion(major: 3, minor: 2, patch: 1)))
+
+		describe("requirementsByDependency") {
+			it("should group dependencies by parent dependency") {
+				let resolvedCartfile = ResolvedCartfile.from(string: validCartfile)
+				let project = Project(directoryURL: URL(string: "file://fake")!)
```

Comment:
There’s a directory — `/var/empty` — which (by implementation) cannot contain real files; would look like `URL(string: "file:///var/empty/fake")!`. Let’s use that.